### ISSUE TITLE
Form errors are not DRY

### ISSUE THREAD

**Issue by uttrasey** [willowtreeapps/react-reform#47]
One of the first things I noticed about the API is the fact the form object has the errors in 2 formats.
- `form.errors: []` and `form.fieldErrors: {}`

I was trying to think of a way of dropping `form.errors` and instead having a function would operate on fieldErrors to give you the output currently provided by `form.errors`

Might be pedantic but it would make the form model really simple. Also having a single source of truth can't be a bad thing.

Wanted to float this to see opinion.

**Comment by kaw2k** [comment]
@uttrasey, good point. We have had this conversation internally before. I thought it was documented, but I couldn't find it. Essentially, there are two problems with forms. 
1. Rendering
2. Business logic

## TL;DR

We currently need to compute the `errors` array in order to pass it to the `<Errors />` component to keep rendering simple. If we are computing it, what is the point of not passing it along and forcing the user to re-compute it. 

Now, for history!

## In the beginning...

When this library first started, `Form` was extremely lightweight. It did not pass down any props to its children, it just serialized inputs, that was it. To get errors to your inputs, you would have to manually pass them down, hence the existence of `fieldErrors`. It looked something like this: 

``` js
React.createClass({
    getInitialState() {
        return { form: getBlankForm() };
    },

    onChange() {
        this.setState({
            form: this.refs.form.serialize()
        });
    },

    render() {
        return <Form ref="form">
            <Input ref="name"
                   onChange={this.onChange}
                   fieldError={this.state.form.fieldErrors.name} />

            <Input ref="age"
                   onChange={this.onChange}
                   fieldError={this.state.form.fieldErrors.age} />

            <ul className="errors">
                {this.state.form.errors.map(e => <li>{e}</li>)}
            </ul>
        </Form>;
    }
});
```

Forms were entirely flat. If you wanted to emulate nested or lists of data, you had to do ungodly things. At the time `fieldErrors` was also a flat object so life was easy. 

## Fast forward to now...

Binding errors manually was a _pain_, it was on our short list of things to fix. Easily displaying concatenated errors was also important. Now we have:

``` js
React.createClass({
    render() {
        return <Form>
            <Input name="name" />
            <Input name="age" />
            <Errors />
        </Form>;
    }
});
```

Much cleaner :smile:. 

## The problem at hand

With all this in mind, we can see that the rendering problem of forms is pretty much solved. The second problem is business logic. How can we use the data from the form to either start an action or produce an error state? 

What of these errors (if any) do we need in an `onSubmit` to further validate our business logic? Most `onSubmit` functions will look something like...

``` js
onSubmit(form) {
if (!form.valid) return;
// Continue...
}
```

We wont even be looking at the errors. With this said, I could _maybe_ see a use case where we might trigger certain actions or analytics of certain fields have errors.

**Comment by uttrasey** [comment]
Update from offline conversation...

It would be nice to combine `fieldErrors` & `errors` into a single error model. We need to think about how `fieldset`, `fieldlist` and `form` level errors would fit in (as currently they wouldn't)

We can always collapse this new model into something that resembles `errors[]` during the clone of `<Errors />` which is looking like the only place it would be needed in that form.

:bread:

**Comment by kaw2k** [comment]
As we test this more, maybe this will evolve some. For now, closing it.
